projects/controllable_dialogue/controllable_seq2seq/controllable_seq2seq.py

```diff
@@ -807,35 +809,43 @@ def beam_search(
         min_n_best=5,
         max_ts=40,
         block_ngram=0,
-        wd_features=[],
-        wd_wts=[],
+        wd_features=None,
+        wd_wts=None,
     ):
-        """ Beam search given the model and Batch
+        """
+        Beam search given the model and Batch.
+
         This function uses model with the following reqs:
+
         - model.encoder takes input returns tuple (enc_out, enc_hidden, attn_mask)
         - model.decoder takes decoder params and returns decoder outputs after attn
         - model.output takes decoder outputs and returns distr over dictionary
 
-        Function arguments:
-        model : nn.Module, here defined in modules.py
-        batch : Batch structure with input and labels
-        beam_size : Size of each beam during the search
-        start : start of sequence token
-        end : end of sequence token
-        pad : padding token
-        min_length : minimum length of the decoded sequence
-        min_n_best : minimum number of completed hypothesis generated from each beam
-        max_ts: the maximum length of the decoded sequence
-        wd_features: list of strings, the WD features to use
-        wd_weights: list of floats, the WD weights to use
-
-        Return:
-        beam_preds_scores : list of tuples (prediction, score) for each sample in Batch
-        n_best_preds_scores : list of n_best list of tuples (prediction, score) for
-                              each sample from Batch
-        beams : list of Beam instances defined in Beam class, can be used for any
-                following postprocessing, e.g. dot logging.
+        :param model: nn.Module, here defined in modules.py
+        :param batch: Batch structure with input and labels
+        :param beam_size: Size of each beam during the search
+        :param start: start of sequence token
+        :param end: end of sequence token
+        :param pad: padding token
+        :param min_length: minimum length of the decoded sequence
+        :param min_n_best: minimum number of completed hypothesis generated
+            from each beam
+        :param max_ts: the maximum length of the decoded sequence
+        :param wd_features: list of strings, the WD features to use
+        :param wd_weights: list of floats, the WD weights to use
+
+        :return:
+            - beam_preds_scores : list of tuples (prediction, score) for each
+              sample in Batch
+            - n_best_preds_scores : list of n_best list of tuples (prediction,
+              score) for each sample from Batch
+            - beams : list of Beam instances defined in Beam class, can be used
+              for any following postprocessing, e.g. dot logging.
         """
+        if wd_features is None:
```

Comment:
what is the point of changing the defaults to None and then immediately converting them to lists?

Reply:
Lists are mutable objects, and default parameters are fixed. That means that if you do any modifications to that list, it changes the default argument for *all future calls*.

Reply:
cuhraaaaaaaazy 😱